libsplinter/src/rest_api/mod.rs

```diff
@@ -330,6 +330,93 @@ impl RequestGuard for Box<dyn RequestGuard> {
     }
 }
 
+/// Guards requests based on a minimum protocol version.
+///
+/// A protocol version is specified via the HTTP header `"X-SplinterProtocolVersion"`.  This header
+/// is a positive integer value.
+#[derive(Clone)]
+pub struct ProtocolVersionRangeGuard {
+    min: u32,
+    max: u32,
+}
+
+impl ProtocolVersionRangeGuard {
+    /// Constructs a new protocol version guard with the given minimum.
+    pub fn new(min: u32, max: u32) -> Self {
+        Self { min, max }
+    }
+}
+
+impl RequestGuard for ProtocolVersionRangeGuard {
+    fn evaluate(&self, req: &HttpRequest) -> Continuation {
+        if let Some(header_value) = req.headers().get("X-SplinterProtocolVersion") {
+            match header_value
+                .to_str()
+                .map_err(|err| {
+                    format!(
+                        "Invalid characters in X-SplinterProtocolVersion header: {}",
+                        err
+                    )
+                })
+                .and_then(|val_str| {
+                    val_str.parse::<u32>().map_err(|_| {
+                        format!("X-SplinterProtocolVersion must be a valid positive integer")
+                    })
+                }) {
+                Err(msg) => Continuation::terminate(
+                    HttpResponse::BadRequest()
+                        .json(json!({
+                            "message": msg,
+                        }))
+                        .into_future(),
+                ),
+                Ok(version) if version < self.min => Continuation::terminate(
+                    HttpResponse::BadRequest()
+                        .json(json!({
+                            "message": format!(
+                                "Client must support protocol version {} or greater.",
+                                self.min,
+                            ),
+                            "requested_protocol": version,
+                            "splinter_protocol": self.max,
+                            "libsplinter_version": format!(
+                                "{}.{}.{}",
+                                env!("CARGO_PKG_VERSION_MAJOR"),
+                                env!("CARGO_PKG_VERSION_MINOR"),
+                                env!("CARGO_PKG_VERSION_PATCH")
+                            )
+                        }))
+                        .into_future(),
+                ),
+                Ok(version) if version > self.max => Continuation::terminate(
+                    HttpResponse::BadRequest()
+                        .json(json!({
+                            "message": format!(
+                                "Client requires a newer protocol than provided: {} > {}",
+                                version,
```

Comment:
This error message is actually a little difficult to make sense of at first; it might be more clear as "Client's requested protocol ({}) is not supported; supported protocols: {}" or something similar. This would also allow you to collapse these two match arms into one.

Reply:
They signify very different problems - one is that your protocol is too old, and you should upgrade your client, the other is that your server is too old to work with the protocol in question, so you have to possibly talk to a sysadmin to fix the issue.